### ISSUE TITLE
optional="True" not respected anymore

### ISSUE THREAD

**Issue by mvdbeek** [galaxyproject/galaxy#6219]
With 18.05 being released planemo is erroring here:

https://travis-ci.org/bardin-lab/readtagger/jobs/383262662#L3179

The relevant part:
```
( findcluster ) > Test-1 ... 2018-05-24 16:15:11,438 INFO  [galaxy.tools.actions.upload_common] tool upload1 created job id 18
2018-05-24 16:15:12,153 INFO  [galaxy.jobs.handler] (18) Job dispatched
2018-05-24 16:15:12,443 INFO  [galaxy.jobs.command_factory] Built script [/tmp/tmpRwXpl7/job_working_directory/000/18/tool_script.sh] for tool command [python '/tmp/tmpRwXpl7/galaxy-dev/tools/data_source/upload.py' /tmp/tmpRwXpl7/galaxy-dev /tmp/tmpRwXpl7/job_working_directory/000/18/registry.xml /tmp/tmpRwXpl7/job_working_directory/000/18/upload_params.json 24:/tmp/tmpRwXpl7/job_working_directory/000/18/dataset_24_files:/tmp/tmpRwXpl7/files/000/dataset_24.dat]
ERROR
( findcluster ) > Test-2 ... 2018-05-24 16:15:19,432 INFO  [galaxy.tools.actions.upload_common] tool upload1 created job id 19
2018-05-24 16:15:20,680 INFO  [galaxy.jobs.handler] (19) Job dispatched
2018-05-24 16:15:20,967 INFO  [galaxy.jobs.command_factory] Built script [/tmp/tmpRwXpl7/job_working_directory/000/19/tool_script.sh] for tool command [python '/tmp/tmpRwXpl7/galaxy-dev/tools/data_source/upload.py' /tmp/tmpRwXpl7/galaxy-dev /tmp/tmpRwXpl7/job_working_directory/000/19/registry.xml /tmp/tmpRwXpl7/job_working_directory/000/19/upload_params.json 25:/tmp/tmpRwXpl7/job_working_directory/000/19/dataset_25_files:/tmp/tmpRwXpl7/files/000/dataset_25.dat]
2018-05-24 16:15:27,111 INFO  [galaxy.tools.actions.upload_common] tool upload1 created job id 20
2018-05-24 16:15:28,018 INFO  [galaxy.jobs.handler] (20) Job dispatched
2018-05-24 16:15:28,544 INFO  [galaxy.jobs.command_factory] Built script [/tmp/tmpRwXpl7/job_working_directory/000/20/tool_script.sh] for tool command [python '/tmp/tmpRwXpl7/galaxy-dev/tools/data_source/upload.py' /tmp/tmpRwXpl7/galaxy-dev /tmp/tmpRwXpl7/job_working_directory/000/20/registry.xml /tmp/tmpRwXpl7/job_working_directory/000/20/upload_params.json 26:/tmp/tmpRwXpl7/job_working_directory/000/20/dataset_26_files:/tmp/tmpRwXpl7/files/000/dataset_26.dat]
ERROR
...
======================================================================
ERROR: ( findcluster ) > Test-1
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmpztusv70q/galaxy-dev/test/functional/test_toolbox.py", line 88, in test_tool
    self.do_it(tool_version=tool_version, test_index=test_index)
  File "/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmpztusv70q/galaxy-dev/test/functional/test_toolbox.py", line 41, in do_it
    verify_tool(tool_id, self.galaxy_interactor, resource_parameters=resource_parameters, test_index=test_index, tool_version=tool_version, register_job_data=register_job_data)
  File "/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmpztusv70q/galaxy-dev/lib/galaxy/tools/verify/interactor.py", line 674, in verify_tool
    raise e
RunToolException: Error creating a job for these tool inputs - Parameter ref_file requires a value, but has no legal values defined., Parameter ref_file requires a value, but has no legal values defined.

======================================================================
ERROR: ( findcluster ) > Test-2
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmpztusv70q/galaxy-dev/test/functional/test_toolbox.py", line 88, in test_tool
    self.do_it(tool_version=tool_version, test_index=test_index)
  File "/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmpztusv70q/galaxy-dev/test/functional/test_toolbox.py", line 41, in do_it
    verify_tool(tool_id, self.galaxy_interactor, resource_parameters=resource_parameters, test_index=test_index, tool_version=tool_version, register_job_data=register_job_data)
  File "/private/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/tmpztusv70q/galaxy-dev/lib/galaxy/tools/verify/interactor.py", line 674, in verify_tool
    raise e
RunToolException: Error creating a job for these tool inputs - Parameter ref_file requires a value, but has no legal values defined.
``` 

`ref_file` is optional (and not provided in Test-1) and defined [here](https://github.com/bardin-lab/readtagger/blob/master/galaxy/macros.xml#L29)

This used to work fine on 18.01 (and still does with `--galaxy_branch release_18.01`)

**Comment by mvdbeek** [comment]
Also this is the upload_params.json:
```json
[
  {
    "to_posix_lines": true,
    "file_type": "fasta",
    "dbkey": "?",
    "auto_decompress": true,
    "run_as_real_user": false,
    "path": "/var/folders/df/6xqpqpcd7h73b6jpx9t6cwhw0000gn/T/upload_file_data_WPI6SD",
    "dataset_id": 3,
    "is_binary": false,
    "link_data_only": "copy_files",
    "uuid": null,
    "name": "reference.fasta",
    "space_to_tab": false,
    "check_content": false,
    "ext": "fasta",
    "purge_source": true,
    "type": "file"
  }
]
```

which looks ok ?

**Comment by mvdbeek** [comment]
OK, this also fails in the UI, so I guess something related to the tool form building is overly aggressive / not respecting the `optional="True"` anymore

![screen shot 2018-05-25 at 11 15 10](https://user-images.githubusercontent.com/6804901/40536611-1318ae88-600d-11e8-907b-26dbeb10e587.png)

**Comment by nsoranzo** [comment]
Maybe you can `git bisect` it if it's easy to reproduce?

**Comment by mvdbeek** [comment]
Yeah, I can point planemo against a local galaxy tree (hopefully)

**Comment by mvdbeek** [comment]
```
git bisect start v18.05 v18.01 --
git bisect run planemo test --galaxy_root /Users/mvandenb/src/galaxy --no_dependency_resolution --no_conda_auto_init --no_cleanup ~/src/readtagger/galaxy/findcluster.xml
```

let's see .. it's running

**Comment by nsoranzo** [comment]
You may also need to add `find . -name *.pyc -delete` before `planemo test`, if there are migrations `.pyc` files with a `.py` the Galaxy started by planemo fails to start.

**Comment by mvdbeek** [comment]
yep, noticed that already. This is the script I'm using:
```
git stash && find . -name \*.pyc -delete && planemo test --galaxy_root /Users/mvandenb/src/galaxy --no_dependency_resolution --no_conda_auto_init --no_cleanup ~/src/readtagger/galaxy/findcluster.xml
```

**Comment by mvdbeek** [comment]
(need to stash or remove the client build files that can conflict)

**Comment by mvdbeek** [comment]
Hmm, it came out at 91e97b60968d6813bd7a28d0d60e778b288ac385

**Comment by mvdbeek** [comment]
That makes sense too, because we're not reaching the 
```
elif value is None:
    if self.optional:
        return None
```

**Comment by mvdbeek** [comment]
Reverting 91e97b60968d6813bd7a28d0d60e778b288ac385 seems like a good option, what do you think @nsoranzo  ?

**Comment by dannon** [comment]
Fixed in #6222
